sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -216,7 +216,25 @@ class SessionCatalog(
     val table = formatTableName(tableDefinition.identifier.table)
     val newTableDefinition = tableDefinition.copy(identifier = TableIdentifier(table, Some(db)))
     requireDbExists(db)
-    externalCatalog.createTable(db, newTableDefinition, ignoreIfExists)
+
+    if (newTableDefinition.tableType == CatalogTableType.EXTERNAL) {
+      // !! HACK ALERT !!
+      //
+      // See https://issues.apache.org/jira/browse/SPARK-15269 for more details about why we have to
+      // set `locationUri` and then remove the directory after creating the external table:
+      val tablePath = defaultTablePath(newTableDefinition.identifier)
+      try {
+        externalCatalog.createTable(
+          db,
+          newTableDefinition.withNewStorage(locationUri = Some(tablePath)),
+          ignoreIfExists)
+      } finally {
+        val path = new Path(tablePath)
+        FileSystem.get(path.toUri, hadoopConf).delete(path, true)
+      }
+    } else {
+      externalCatalog.createTable(db, newTableDefinition, ignoreIfExists)
+    }
```

Comment:
I am wondering if we should worry about the case where the `defaultTablePath` happens to be the same as the user-specified path for creating external table that contains real data. It may delete the external table data. For example:
```
create table t10 (c1 int) using parquet options(path '/Users/xinwu/spark/spark-warehouse/t10');
insert into t10 values (1);
drop table t10;
create table t10 (c1 int) using parquet options(path '/Users/xinwu/spark/spark-warehouse/t10');
```
In the above case, my metastore warehouse dir is `/Users/xinwu/spark/spark-warehouse', so the `defaultTablePath` will return `'/Users/xinwu/spark/spark-warehouse/t10'`.  Now, upon the creation of the 2nd table, the data path will be deleted, right? Maybe this is a corner case that we may not worry about. but I thought I should bring it up. 

Another observation is that in a hive-compatible case (above case), `createDataSourceTabes` set the `locationURI` with the user specified path, but will be overridden by the above code. Then, users will not be able to query anything back from hive shell, unless users don't expect to see same results from hive shell for hive-compatible tables. I am not sure about the semantic of hive-compatible datasource table. Will this be a problem? Thanks!